File: src/Domain/Common/RichText.cs
using System.Globalization;
using System.Text;

namespace ErpApp.Domain.Common;

/// <summary>
/// Phase 39 -- the one place this codebase turns arbitrary HTML into a rich-text field it is willing
/// to store, print and email.
///
/// <para><b>Why a sanitiser is not optional here.</b> The reference product's editor is TinyMCE
/// 7.1.1 with a <c>code</c> ("Source code") button and <i>no</i> <c>valid_elements</c> configured --
/// both read off the live tenant on 2026-09-13. A user can therefore type raw markup by hand, and
/// what they type is rendered later on a document detail page, in a PDF a customer receives, and in
/// an HTML email body. Three render targets, one of which is a browser, means an unsanitised field
/// is a stored-XSS hole by construction.</para>
///
/// <para><b>The mechanism is re-emission, not filtering.</b> Nothing here asks "is this tag safe?"
/// or "is this attribute safe?" and passes the answer through. The input is parsed into
/// <see cref="RichTextNode"/> -- a tree with no attribute slot except an <i>enum</i>
/// (<see cref="RichTextAlignment"/>) -- and the output is generated from that tree out of string
/// constants this file owns, with every text run escaped on the way. The security property that
/// follows is worth stating precisely, because it is what makes a hand-written parser acceptable at
/// all: <b>a bug in the tokenizer can produce wrong formatting, and cannot produce an attribute, a
/// tag name or a URL that came from the input.</b> The only user bytes that survive are text-node
/// characters, and those are escaped.</para>
///
/// <para><b>Where it runs.</b> On write, in the Domain -- <c>Invoice.SetTerms</c> and its four
/// siblings, <c>CustomTemplate</c>, <c>EmailTemplate</c> and the send path all call
/// <see cref="Sanitize"/> before storing. Sanitising on read would mean every future read path has
/// to remember (phase-35a's lesson is that read paths are exactly what a sweep forgets), and
/// sanitising in a handler would leave a Domain caller able to bypass it. Stored content is
/// therefore already safe, and <see cref="Sanitize"/> is idempotent so re-saving never
/// double-escapes.</para>
/// </summary>
public static class RichText
{
    /// <summary>
    /// Upper bound on the <i>sanitised</i> output, in characters. Generous: a page of terms and
    /// conditions with markup runs to a few thousand. It exists so a paste of an entire web page
    /// cannot put an unbounded string into a column, not to police prose length.
    /// </summary>
    public const int MaxLength = 64_000;

    /// <summary>
    /// Canonical, safe HTML for <paramref name="html"/>, or null when it carries no content.
    ///
    /// <para>Idempotent: <c>Sanitize(Sanitize(x)) == Sanitize(x)</c> for every input. That matters
    /// in practice rather than in theory -- a document is loaded, shown in the editor and saved
    /// again on every edit, so a non-idempotent sanitiser would visibly rot a field over a few
    /// saves.</para>
    /// </summary>
    public static string? Sanitize(string? html)
    {
        var blocks = Parse(html);

        if (blocks.Count == 0)
        {
            return null;
        }

        var builder = new StringBuilder();
        WriteBlocks(builder, blocks);
        var result = builder.ToString();

        return result.Length == 0 ? null : result;
    }

    /// <summary>
    /// The tree behind <paramref name="html"/>: a list of block-level nodes. Public because the
    /// print pipeline folds over the same tree the stored HTML is emitted from -- that shared fold
    /// is what keeps the screen and the PDF from disagreeing.
    /// </summary>
    public static IReadOnlyList<RichTextNode> Parse(string? html)
    {
        if (string.IsNullOrWhiteSpace(html))
        {
            return [];
        }

        var raw = RawParser.Parse(html);
        var blocks = Normalizer.ToBlocks(raw);

        return Trim(blocks);
    }

    /// <summary>
    /// The field's text with all markup removed, for a spreadsheet cell, a list preview or a search
    /// index -- anywhere a single string is wanted and markup would be noise. List items keep a
    /// bullet so a flattened list still reads as one.
    /// </summary>
    public static string ToPlainText(string? html)
    {
        var lines = new List<string>();
        WritePlainBlocks(Parse(html), lines);

        return string.Join("\n", lines).Trim();
    }

    /// <summary>True when the field would store nothing -- markup with no text, whitespace, or null.
    /// A validator wanting "terms were actually entered" asks this, never <c>string.IsNullOrEmpty</c>
    /// on the raw input, which <c>&lt;p&gt;&lt;br&gt;&lt;/p&gt;</c> would pass.</summary>
    public static bool IsEmpty(string? html) => Sanitize(html) is null;

    /// <summary>
    /// Wraps plain text as rich text, escaping it and turning blank-line-separated runs into
    /// paragraphs and single newlines into breaks.
    ///
    /// <para>This is how the phase-39 migration converts the plain-text <c>Terms</c> that phases 27b
    /// and 20d stored. Detecting "is this row HTML or not?" at render time would have left two
    /// representations in one column forever; converting once means there is exactly one.</para>
    /// </summary>
    public static string? FromPlainText(string? text)
    {
        if (string.IsNullOrWhiteSpace(text))
        {
            return null;
        }

        var normalized = text.Replace("\r\n", "\n").Replace('\r', '\n');
        var paragraphs = normalized.Split("\n\n", StringSplitOptions.None);
        var builder = new StringBuilder();

        foreach (var paragraph in paragraphs)
        {
            if (paragraph.Trim().Length == 0)
            {
                continue;
            }

            builder.Append("<p>");
            var first = true;

            foreach (var line in paragraph.Split('\n'))
            {
                if (!first)
                {
                    builder.Append("<br />");
                }

                Escape(builder, line);
                first = false;
            }

            builder.Append("</p>");
        }

        return builder.Length == 0 ? null : builder.ToString();
    }

    // --- emitting -----------------------------------------------------------------------------

    private static void WriteBlocks(StringBuilder builder, IReadOnlyList<RichTextNode> blocks)
    {
        foreach (var block in blocks)
        {
            switch (block)
            {
                case RichTextNode.Paragraph paragraph:
                    builder.Append(paragraph.Alignment == RichTextAlignment.Start
                        ? "<p>"
                        : $"<p style=\"text-align:{AlignmentCss(paragraph.Alignment)}\">");
                    WriteInline(builder, paragraph.Children);
                    builder.Append("</p>");
                    break;

                case RichTextNode.List list:
                    builder.Append(list.Ordered ? "<ol>" : "<ul>");
                    foreach (var item in list.Items)
                    {
                        builder.Append("<li>");
                        WriteInline(builder, item.Children);
                        builder.Append("</li>");
                    }

                    builder.Append(list.Ordered ? "</ol>" : "</ul>");
                    break;

                case RichTextNode.Rule:
                    builder.Append("<hr />");
                    break;
            }
        }
    }

    private static void WriteInline(StringBuilder builder, IReadOnlyList<RichTextNode> nodes)
    {
        foreach (var node in nodes)
        {
            switch (node)
            {
                case RichTextNode.Text text:
                    Escape(builder, text.Value);
                    break;

                case RichTextNode.LineBreak:
                    builder.Append("<br />");
                    break;

                case RichTextNode.Emphasis emphasis:
                    var tag = EmphasisTag(emphasis.Style);
                    builder.Append('<').Append(tag).Append('>');
                    WriteInline(builder, emphasis.Children);
                    builder.Append("</").Append(tag).Append('>');
                    break;
            }
        }
    }

    /// <summary>The four constants that are the entire tag vocabulary of an emphasis run.</summary>
    private static string EmphasisTag(RichTextStyle style) => style switch
    {
        RichTextStyle.Bold => "strong",
        RichTextStyle.Italic => "em",
        RichTextStyle.Underline => "u",
        _ => "s",
    };

    private static string AlignmentCss(RichTextAlignment alignment) => alignment switch
    {
        RichTextAlignment.Center => "center",
        RichTextAlignment.End => "right",
        RichTextAlignment.Justify => "justify",
        _ => "left",
    };

    private static void Escape(StringBuilder builder, string value)
    {
        foreach (var c in value)
        {
            switch (c)
            {
                case '&': builder.Append("&amp;"); break;
                case '<': builder.Append("&lt;"); break;
                case '>': builder.Append("&gt;"); break;
                case '"': builder.Append("&quot;"); break;
                case '\'': builder.Append("&#39;"); break;
                default: builder.Append(c); break;
            }
        }
    }

    // --- flattening ---------------------------------------------------------------------------

    private static void WritePlainBlocks(IReadOnlyList<RichTextNode> blocks, List<string> lines)
    {
        foreach (var block in blocks)
        {
            switch (block)
            {
                case RichTextNode.Paragraph paragraph:
                    lines.AddRange(PlainLines(paragraph.Children));
                    break;

                case RichTextNode.List list:
                    var index = 1;
                    foreach (var item in list.Items)
                    {
                        var prefix = list.Ordered
                            ? $"{index.ToString(CultureInfo.InvariantCulture)}. "
                            : "• ";
                        var itemLines = PlainLines(item.Children);
                        lines.Add(prefix + (itemLines.Count > 0 ? itemLines[0] : string.Empty));
                        for (var i = 1; i < itemLines.Count; i++)
                        {
                            lines.Add(new string(' ', prefix.Length) + itemLines[i]);
                        }

                        index++;
                    }

                    break;

                case RichTextNode.Rule:
                    lines.Add("---");
                    break;
            }
        }
    }

    private static List<string> PlainLines(IReadOnlyList<RichTextNode> nodes)
    {
        var lines = new List<string> { string.Empty };

        void Walk(IReadOnlyList<RichTextNode> children)
        {
            foreach (var node in children)
            {
                switch (node)
                {
                    case RichTextNode.Text text:
                        lines[^1] += text.Value;
                        break;
                    case RichTextNode.LineBreak:
                        lines.Add(string.Empty);
                        break;
                    case RichTextNode.Emphasis emphasis:
                        Walk(emphasis.Children);
                        break;
                }
            }
        }

        Walk(nodes);

        return lines;
    }

    // --- trimming -----------------------------------------------------------------------------

    /// <summary>
    /// Drops empty paragraphs at the two ends only. The editor emits a trailing
    /// <c>&lt;p&gt;&lt;br&gt;&lt;/p&gt;</c> almost every time; an empty paragraph <i>between</i> two
    /// filled ones is a blank line somebody typed on purpose, so it stays.
    /// </summary>
    private static List<RichTextNode> Trim(List<RichTextNode> blocks)
    {
        static bool IsBlank(RichTextNode node) =>
            node is RichTextNode.Paragraph { Children.Count: 0 };

        var start = 0;
        var end = blocks.Count;

        while (start < end && IsBlank(blocks[start]))
        {
            start++;
        }

        while (end > start && IsBlank(blocks[end - 1]))
        {
            end--;
        }

        return blocks.GetRange(start, end - start);
    }
}

File: web/src/app/shared/rich-text/rich-text.ts
/**
 * Phase 39 — the client half of `Domain/Common/RichText.cs`.
 *
 * <b>Why there are two.</b> The server's copy is the one that decides what gets stored, and it would
 * be sufficient on its own for safety. This one exists so that what the user sees while typing is
 * what will be stored: without it, pasting a coloured table from Word shows a coloured table, the
 * save silently returns a plain paragraph, and the field appears to have eaten the content. The
 * server remains authoritative — nothing here is trusted by anything.
 *
 * <b>How the two are kept in step.</b> `rich-text-cases.json` sits beside this file and is read by
 * both suites: by `rich-text.spec.ts` here, and by `RichTextSharedCasesTests` in Domain.UnitTests,
 * which links the same file in as an embedded resource. That is phase-26b's rule for `BsCalendar`
 * and its `bs-date.ts` twin — port it, never retype it, and pin both boundaries — applied to the
 * second pair of twins in this codebase.
 *
 * <b>The same security property as the server's.</b> This is re-emission, not filtering: the input
 * is parsed with `DOMParser` (which builds a detached document — it runs no script and fetches no
 * resource), walked, and written back out of string constants this file owns, with text escaped.
 * No attribute the user supplied is ever copied; the one attribute emitted is a `text-align` chosen
 * from four constants.
 */

/** The four character styles, matching `RichTextStyle`. */
const BOLD = 1;
const ITALIC = 2;
const UNDERLINE = 4;
const STRIKE = 8;

type Alignment = 'left' | 'center' | 'right' | 'justify';

/** Elements whose content is discarded with them, matching `RawParser.Opaque`. */
const OPAQUE = new Set([
  'SCRIPT', 'STYLE', 'IFRAME', 'OBJECT', 'EMBED', 'SVG', 'MATH', 'NOSCRIPT', 'TEMPLATE',
  'TEXTAREA', 'TITLE', 'HEAD', 'FRAME', 'FRAMESET', 'APPLET',
]);

/** Elements that end the paragraph in progress, matching `Normalizer.Blocks`. */
const BLOCKS = new Set([
  'P', 'DIV', 'BLOCKQUOTE', 'PRE', 'SECTION', 'ARTICLE', 'HEADER', 'FOOTER', 'MAIN', 'ASIDE',
  'FIGURE', 'FIGCAPTION', 'ADDRESS', 'DL', 'DD', 'DT', 'TR', 'CAPTION', 'FORM', 'FIELDSET',
  'H1', 'H2', 'H3', 'H4', 'H5', 'H6',
]);

const HEADINGS = new Set(['H1', 'H2', 'H3', 'H4', 'H5', 'H6']);

const EMPHASES: Readonly<Record<string, number>> = {
  B: BOLD, STRONG: BOLD,
  I: ITALIC, EM: ITALIC,
  U: UNDERLINE, INS: UNDERLINE,
  S: STRIKE, STRIKE: STRIKE, DEL: STRIKE,
};

/** One paragraph or list item under construction. */
interface Flow {
  readonly parts: string[];
  visible: boolean;
  breaks: boolean;
}

/** Canonical, safe HTML, or '' when the input carries no content. Mirrors `RichText.Sanitize`. */
export function sanitizeRichText(html: string | null | undefined): string {
  if (!html || html.trim().length === 0) {
    return '';
  }

  const parsed = new DOMParser().parseFromString(html, 'text/html');
  const writer = new BlockWriter();
  walk(parsed.body, writer, 'left', 0);
  writer.flush();

  return writer.finish();
}

/** True when the field would store nothing. Mirrors `RichText.IsEmpty`. */
export function isRichTextEmpty(html: string | null | undefined): boolean {
  return sanitizeRichText(html).length === 0;
}

/** The field's text with markup removed, for a preview line or a title attribute. */
export function richTextToPlainText(html: string | null | undefined): string {
  const parsed = new DOMParser().parseFromString(sanitizeRichText(html), 'text/html');

  return (parsed.body.textContent ?? '').trim();
}

/**
 * Wraps plain text as rich text, escaping it. Mirrors `RichText.FromPlainText` — used when a caller
 * has plain text and needs to hand it to the editor.
 */
export function richTextFromPlainText(text: string | null | undefined): string {
  if (!text || text.trim().length === 0) {
    return '';
  }

  return text
    .replace(/\r\n/g, '\n')
    .replace(/\r/g, '\n')
    .split('\n\n')
    .filter((paragraph) => paragraph.trim().length > 0)
    .map((paragraph) => `<p>${paragraph.split('\n').map(escape).join('<br />')}</p>`)
    .join('');
}

// --- the walk -------------------------------------------------------------------------------

class BlockWriter {
  private readonly blocks: string[] = [];

  private flow: Flow = { parts: [], visible: false, breaks: false };

  alignment: Alignment = 'left';

  write(markup: string, visible: boolean): void {
    this.flow.parts.push(markup);
    this.flow.visible ||= visible;
  }

  writeBreak(): void {
    this.flow.parts.push('<br />');
    this.flow.breaks = true;
  }

  addBlock(markup: string): void {
    this.blocks.push(markup);
  }

  /** Ends the paragraph in progress. A paragraph of only whitespace is dropped; one holding only
   * breaks becomes the empty paragraph that `finish` may then trim. */
  flush(): void {
    const { parts, visible, breaks } = this.flow;
    this.flow = { parts: [], visible: false, breaks: false };

    if (parts.length === 0) {
      return;
    }

    if (!visible) {
      if (breaks) {
        this.blocks.push('<p></p>');
      }

      return;
    }

    const body = parts.join('').replace(/^(\s|&nbsp;)+/, '').replace(/(\s|&nbsp;)+$/, '');

    if (body.length === 0) {
      return;
    }

    this.blocks.push(this.alignment === 'left' ? `<p>${body}</p>` : `<p style="text-align:${this.alignment}">${body}</p>`);
  }

  /** Trims empty paragraphs at the two ends only — the editor emits a trailing one almost every
   * time, while one between two filled paragraphs is a blank line somebody typed. */
  finish(): string {
    let start = 0;
    let end = this.blocks.length;

    while (start < end && this.blocks[start] === '<p></p>') {
      start++;
    }

    while (end > start && this.blocks[end - 1] === '<p></p>') {
      end--;
    }

    return this.blocks.slice(start, end).join('');
  }

  /** Renders one list item's inline content, reusing the same flow machinery. */
  static inline(node: Node, alignment: Alignment, style: number): string {
    const inner = new BlockWriter();
    inner.alignment = alignment;
    walkChildren(node, inner, alignment, style);
    inner.flush();

    return inner.blocks
      .map((block) => block.replace(/^<p[^>]*>/, '').replace(/<\/p>$/, ''))
      .filter((x) => x.length > 0)
      .join('<br />');
  }
}

function walk(node: Node, writer: BlockWriter, alignment: Alignment, style: number): void {
  if (node.nodeType === Node.TEXT_NODE) {
    const value = (node.nodeValue ?? '').replace(/\s+/g, ' ');

    if (value.length > 0) {
      writer.write(styled(escape(value), style), value.trim().length > 0);
    }

    return;
  }

  if (node.nodeType !== Node.ELEMENT_NODE) {
    return;
  }

  const element = node as Element;
  const name = element.tagName.toUpperCase();

  if (OPAQUE.has(name)) {
    return;
  }

  const inherited = alignmentOf(element) ?? alignment;

  if (name === 'BR') {
    writer.writeBreak();

    return;
  }

  if (name === 'HR') {
    writer.flush();
    writer.addBlock('<hr />');

    return;
  }

  if (name === 'UL' || name === 'OL') {
    writer.flush();
    const tag = name === 'OL' ? 'ol' : 'ul';
    const items = Array.from(element.children)
      .filter((child) => child.tagName.toUpperCase() === 'LI')
      .map((item) => `<li>${BlockWriter.inline(item, alignmentOf(item) ?? inherited, style)}</li>`);

    if (items.length > 0) {
      writer.addBlock(`<${tag}>${items.join('')}</${tag}>`);
    }

    return;
  }

  const emphasis = EMPHASES[name];

  if (emphasis !== undefined) {
    walkChildren(element, writer, inherited, style | emphasis);

    return;
  }

  if (BLOCKS.has(name)) {
    writer.flush();
    const previous = writer.alignment;
    writer.alignment = inherited;
    walkChildren(element, writer, inherited, HEADINGS.has(name) ? style | BOLD : style);
    writer.flush();
    writer.alignment = previous;

    return;
  }

  if (name === 'TD' || name === 'TH') {
    walkChildren(element, writer, inherited, style);
    writer.write(' ', false);

    return;
  }

  if (name === 'LI') {
    // A list item outside any list: a paragraph, rather than text thrown away.
    writer.flush();
    walkChildren(element, writer, inherited, style);
    writer.flush();

    return;
  }

  // Everything else unwraps — span, font, a, img, table, tbody, unknown tags. See the file comment.
  walkChildren(element, writer, inherited, style);
}

function walkChildren(node: Node, writer: BlockWriter, alignment: Alignment, style: number): void {
  node.childNodes.forEach((child) => walk(child, writer, alignment, style));
}

/** Wraps one escaped run in a tag per set flag, outermost first — the order the server emits. */
function styled(markup: string, style: number): string {
  let result = markup;

  if (style & STRIKE) {
    result = `<s>${result}</s>`;
  }

  if (style & UNDERLINE) {
    result = `<u>${result}</u>`;
  }

  if (style & ITALIC) {
    result = `<em>${result}</em>`;
  }

  if (style & BOLD) {
    result = `<strong>${result}</strong>`;
  }

  return result;
}

/**
 * Reduces a declared alignment to one of four constants, or null. The value is matched and then
 * discarded: what reaches the output is the constant that matched, never the string.
 */
function alignmentOf(element: Element): Alignment | null {
  const declared =
    (element as HTMLElement).style?.textAlign ||
    element.getAttribute('align') ||
    '';

  switch (declared.trim().toLowerCase()) {
    case 'center':
      return 'center';
    case 'right':
    case 'end':
      return 'right';
    case 'justify':
      return 'justify';
    case 'left':
    case 'start':
      return 'left';
    default:
      return null;
  }
}

function escape(value: string): string {
  return value
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;');
}
